### PR TITLE
Use distinct signal to wait to cancel the submission

### LINKED ISSUE #9298: [Bug] [Flaky test]: CanceledTasksDoNotLogMSB4181 unit test is unreliable
### Issue Description

Unit test `Microsoft.Build.UnitTests.BackEnd.TaskBuilder_Tests.CanceledTasksDoNotLogMSB4181` failed on a PR build.

```
Shouldly.ShouldAssertException : logger.WarningCount
    should be
1
    but was
0
```

The test includes a `Thread.Sleep` call.

### Steps to Reproduce

The issue was found when running the set of unit tests during a PR build but it appears to be a threading race condition and the test doesn't consistently break.

### Expected Behavior

The test should pass for the PR build where it failed.

### Actual Behavior

The test failed for Windows Core while passing for the other type of builds within the PR build.

### Analysis

Related to:
#9277
#9100

A search in the repo shows that `Thread.Sleep` appears 163 times in 32 files. About 22 of the files appear to be unit tests or test support.

### Versions & Configurations

_No response_

### PR DESCRIPTION
**Pull request by GangWang01**
Fixes #9298

### Context
Waiting using `Thread.Sleep(500)` in the test is unreliable.

### Changes Made
Wait until the specified log appears rather than `Thread.Sleep(500)`.

### CODE REVIEW

File: src/Build.UnitTests/BackEnd/TaskBuilder_Tests.cs [hunk 1]
```diff
@@ -168,8 +170,18 @@ public void CanceledTasksDoNotLogMSB4181()
                 BuildRequestData data = new BuildRequestData(project.CreateProjectInstance(), new string[] { "test" }, collection.HostServices);
                 manager.BeginBuild(_parameters);
                 BuildSubmission asyncResult = manager.PendBuildRequest(data);
+                string unescapedSleepCommand = sleepCommand.Replace("&quot;", "\"").Replace("&gt;", ">");
+                Func<bool> isSleepCommandExecuted = () => logger.AllBuildEvents.Any(a => unescapedSleepCommand.Equals(a.Message));
```
Comment (JanKrivanek) [inline src/Build.UnitTests/BackEnd/TaskBuilder_Tests.cs:174]: While accomplishing the purpose, it feels superfluous to iterate all the log events over and over again (I guess it risk getting the collection modified exception as well).

It should be easy to check only the new events (if any) on each try

Comment (JanKrivanek) [reply src/Build.UnitTests/BackEnd/TaskBuilder_Tests.cs:174]: It seems [`AdditionalHandlers`](https://github.com/dotnet/msbuild/blob/f4fa6bde775a3f7cbb2bb90a349ee5fc759114f3/src/Shared/UnitTests/MockLogger.cs#L248C53-L248C71) can be leveraged here - by registering a handler that would set some waithandle upon receiving the required log entry

**Review by JanKrivanek — CHANGES_REQUESTED** [review]
Looks as a good way to get rid of some of the pesky flakyness - great!

I'd love to see the test to not iterate the log events repetitively. We can use MockLogger to avoid the additional thread/task and 'wake-check-sleep' pattern here

**Review by JanKrivanek — CHANGES_REQUESTED** [review]

**Review by JanKrivanek — APPROVED** [review]
Thank you!

**Review by YuliiaKovalova — APPROVED** [review]

**Pull request merged**